source/math/Quaternion.ooc

```diff
@@ -91,7 +107,79 @@ Quaternion: cover {
 		else
 			roll = asin((upYawPitch scalarProduct(upRotated) * upYawPitch z - upRotated z) / rightYawPitch z)
 
-		FloatRotation3D new(pitch, -yaw, roll)
+		FloatRotation3D new (pitch, -yaw, roll)
+	}
+	norm: Float { get { (this real squared() + (this imaginary norm) squared()) sqrt() } }
+	rotation: Float { get { 2.0f * (this calculateLogarithm() imaginary) norm } }
+	rotationX: Float {
+		get {
+			result: Float
+			value := this w * this y - this z * this x
+			if ((value abs() - 0.5f) abs() < this precision)
+				result = 0.0f
+			else
+				result = (2.0f * (this w * this x + this y * this z)) atan2(1.0f - 2.0f * (this x squared() + this y squared()))
+			result
+		}
+	}
+	rotationY: Float {
+		get {
+			result: Float
+			value := this w * this y - this z * this x
+			if ((value abs() - 0.5f) abs() > this precision)
+				result = ((2.0f * (this w * this y - this z * this x)) clamp(-1, 1)) asin()
+			else
+				result = Float sign(value) * (Float pi / 2.0f)
+			result
+		}
+	}
+	rotationZ: Float {
+		get {
+			result: Float
+			value := this w * this y - this z * this x
+			if ((value abs() - 0.5f) abs() < this precision)
+				result = 2.0f * (this z atan2(this w))
+			else
+				result = (2.0f * (this w * this z + this x * this y)) atan2(1.0f - 2.0f * (this y squared() + this z squared()))
+			result
+		}
+	}	
+	direction: FloatPoint3D {
+		get {
+			quaternionLogarithm := this calculateLogarithm()
+			quaternionLogarithm imaginary / quaternionLogarithm imaginary norm
+		}
+	}
+	distance: func (other: Quaternion) -> Float {
+		(this - other) norm
+	}
+	calculateLogarithm: func -> Quaternion {
+		result := Quaternion new()
+		norm := this norm
+		point3DNorm := this imaginary norm
+		if (point3DNorm != 0) {
+			result real = norm log()
+			result imaginary = (this imaginary / point3DNorm) * ((this real / norm) acos())
+		}
+		else {
+			result real = norm
+			result imaginary = FloatPoint3D new()
+		}
+		result
+	}
+	calculateExponential: func -> Quaternion {
```

Comment:
OK, so - 
 * if you need an argument, it's a function. 
 * if 
     * you don't need an argument and
     * you don't affect the original instance,

    it's a property.